### PR TITLE
feat: Argocd application noncascade delete for cluster unreachable state

### PR DESCRIPTION
**Pull request by Ash-exp**
# Description

Deleting an app deployed with Argocd for the cluster which is no longer accessible gets stuck in partial delete state. For this state introduced non-cascade delete APIs and integrated them to dashboard.

Fixes #https://github.com/devtron-labs/devtron/issues/3304
Ticket [AB 2821](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/OSS%20adoption/Stories/?workitem=2821)
Figma ProtoType [Link](https://www.figma.com/proto/z5oh20tbYNDZkDQGrSYgol/Global-Configurations?page-id=4456%3A169281&type=design&node-id=4457-169825&viewport=761%2C-1604%2C0.44&scaling=min-zoom&starting-point-node-id=4457%3A169825)
TechDoc [Link](https://docs.google.com/document/d/1RNin5ob-5mayKX4jfwGu9sbuDoETddBtAjDBya-13y4/edit?usp=sharing)
Backend PR [Link](https://github.com/devtron-labs/devtron/pull/3339)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Helm App (ArgoCD Deployed)
- [x] Devtron App (ArgoCD Deployed)
- [x] Helm App (Helm Deployed)
- [x] Devtron App (Helm Deployed)

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/details/appDetails/appDetails.service.ts [hunk 1]
```diff
@@ -16,6 +17,23 @@ export function isDatasourceHealthy(datasourceId: number | string) {
     return fetchWithFullRoute(URL, 'GET')
 }
 
+export function getClusterConnectionStatus(envId: number) : Promise<ClusterConnectionResponse> {
```
Comment (milindk8) [inline src/components/app/details/appDetails/appDetails.service.ts:20]: format code

File: src/components/charts/modal/DeployChart.tsx [hunk 2]
```diff
@@ -718,7 +738,7 @@ const DeployChart: React.FC<DeployChartProps> = ({
                 {confirmation && (
                     <DeleteDialog
                         title={`Delete '${originalName}' ?`}
-                        delete={() => handleDelete(false)}
+                        delete={() => handleDelete(DELETE_ACTION.DELETE)}
```
Comment (milindk8) [inline src/components/charts/modal/DeployChart.tsx:741]: remove inline function

File: src/components/charts/modal/DeployChart.tsx [hunk 3]
```diff
@@ -730,14 +750,21 @@ const DeployChart: React.FC<DeployChartProps> = ({
                 {showForceDeleteDialog && (
                     <ForceDeleteDialog
                         forceDeleteDialogTitle={forceDeleteDialogTitle}
-                        onClickDelete={() => handleDelete(true)}
+                        onClickDelete={() => handleDelete(DELETE_ACTION.FORCE_DELETE)}
```
Comment (milindk8) [inline src/components/charts/modal/DeployChart.tsx:753]: same here

File: src/components/v2/appDetails/SyncError.component.tsx [hunk 4]
```diff
@@ -71,6 +141,23 @@ const SyncErrorComponent: React.FC<SyncErrorType> = ({ appStreamData, showApplic
...
+                            <tr>
+                                <td className="pb-8 min-width">Cluster is not reachable</td>
+                                <td className="pl-24 pb-8">
+                                    {`The underlying resources cannot be deleted as the cluster${
+                                        clusterName ? ` '${clusterName}'` : ''
+                                    } is not
+                                    reachable at the moment.`}
+                                    <span
+                                        className="pointer ml-8 cb-5"
+                                        onClick={() => showNonCascadeDeleteDialog(true)}
```
Comment (milindk8) [inline src/components/v2/appDetails/SyncError.component.tsx:154]: remove inline

File: src/components/v2/appDetails/SyncError.component.tsx [hunk 5]
```diff
@@ -88,6 +175,21 @@ const SyncErrorComponent: React.FC<SyncErrorType> = ({ appStreamData, showApplic
                     </tbody>
                 </table>
             )}
+            {forceDeleteDialog && (
+                <ForceDeleteDialog
+                    forceDeleteDialogTitle={forceDeleteDialogTitle}
+                    onClickDelete={() => nonCascadeDeleteArgoCDApp(true)}
```
Comment (milindk8) [inline src/components/v2/appDetails/SyncError.component.tsx:181]: here too

File: src/components/v2/values/chartValuesDiff/ChartValuesView.component.tsx [hunk 6]
```diff
@@ -297,11 +298,14 @@ export const DeleteChartDialog = ({
...
 }: DeleteChartDialogProps) => {
+    const closeConfirmation = () => {
+        toggleConfirmation(false)
+    }
     return (
         <DeleteDialog
             title={`Delete '${appName}' ?`}
-            delete={() => handleDelete(false)}
-            closeDelete={toggleConfirmation}
+            delete={() => handleDelete(DELETE_ACTION.DELETE)}
```
Comment (milindk8) [inline src/components/v2/values/chartValuesDiff/ChartValuesView.component.tsx:307]: remove inline

File: src/components/app/details/appDetails/AppDetails.tsx [hunk 7]
```diff
@@ -499,6 +522,8 @@ export const Details: React.FC<DetailsType> = ({
                         />
                     )}
                 </>
+            ) : (
+                <div className="mb-9"></div>
```
[the remark this answers is not in the document]
Comment (Ash-exp) [reply src/components/app/details/appDetails/AppDetails.tsx:526]: Currently the Resource tree component is getting margin bottom from app metrics component. But in Partial delete state we don't show any of them. this is to create margin leveraging the mentioned condition.

File: src/components/app/details/appDetails/appDetails.service.ts (hunk 1 above)
Comment (Ash-exp) [reply src/components/app/details/appDetails/appDetails.service.ts:20]: implemented 👍

File: src/components/charts/modal/DeployChart.tsx (hunk 2 above)
Comment (Ash-exp) [reply src/components/charts/modal/DeployChart.tsx]: implemented 👍

File: src/components/charts/modal/DeployChart.tsx (hunk 3 above)
Comment (Ash-exp) [reply src/components/charts/modal/DeployChart.tsx]: implemented 👍

File: src/components/v2/appDetails/SyncError.component.tsx (hunk 5 above)
Comment (Ash-exp) [reply src/components/v2/appDetails/SyncError.component.tsx]: implemented 👍

File: src/components/app/details/appDetails/AppDetails.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/app/details/appDetails/AppDetails.tsx:530]: we can write this as <div className="mb-9" />

File: src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import { ConfirmationDialog } from '@devtron-labs/devtron-fe-common-lib'
+import warningIconSrc from '../../../assets/icons/ic-warning-y5.svg'
+import { BUTTON_TEXT, NONCASCADE_DELETE_DIALOG_INTERNAL_MESSAGE } from '../../../config/constantMessaging'
+
+interface ClusrerNotReachableDialogType {
```
Comment (vivek-devtron) [inline src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx:6]: Move this to type

File: src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    clusterName: string
+    onClickCancel: () => void
+    onClickDelete: () => void
+}
+function ClusrerNotReachableDialog({ clusterName, onClickCancel, onClickDelete }: ClusrerNotReachableDialogType) {
+    const clusterNameString:string = clusterName ? ` '${clusterName}'` : ''
+    return (
+        <ConfirmationDialog>
+            <ConfirmationDialog.Icon src={warningIconSrc} />
+            <ConfirmationDialog.Body title={`The cluster${clusterNameString} is not reachable`} />
```
Comment (vivek-devtron) [inline src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx:16]: Should there be space after cluster?

File: src/components/cdPipeline/CDPipeline.tsx [hunk 10]
```diff
@@ -70,6 +70,7 @@ import {
     TOAST_INFO,
     CONFIGMAPS_SECRETS,
 } from '../../config/constantMessaging'
+import ClusrerNotReachableDialog from '../common/ClusterNotReachableDailog/ClusterNotReachableDialog'
```
Comment (vivek-devtron) [inline src/components/cdPipeline/CDPipeline.tsx:73]: Please fix the typo

File: src/components/cdPipeline/CDPipeline.tsx [hunk 11]
```diff
@@ -737,6 +754,28 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
             })
     }
 
+    handleDeletePipeline = (deleteAction: string) => {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/CDPipeline.tsx:757]: this should be of DELETE_ACTION type

File: src/components/charts/discoverChartDetail/ChartDeploymentList.tsx [hunk 12]
```diff
@@ -88,23 +93,27 @@ export function DeploymentRow({ installedAppId, appName, status, environmentId,
         }
     }
 
-    async function handleDelete(force) {
+    async function handleDelete(deleteAction: string) {
```
Comment (vivek-devtron) [inline src/components/charts/discoverChartDetail/ChartDeploymentList.tsx:96]: this should be of DELETE_ACTION type

File: src/components/charts/modal/DeployChart.tsx [hunk 13]
```diff
@@ -35,6 +35,7 @@ import { mainContext } from '../../common/navigation/NavigationRoutes'
 import HyperionEnvironmentSelect from '../../hyperion/EnvironmentSelect'
 import { getAppId } from '../../v2/appDetails/k8Resource/nodeDetail/nodeDetail.api'
 import { DeploymentAppType } from '../../v2/values/chartValuesDiff/ChartValuesView.type'
+import ClusrerNotReachableDialog from '../../common/ClusterNotReachableDailog/ClusterNotReachableDialog'
```
Comment (vivek-devtron) [inline src/components/charts/modal/DeployChart.tsx:38]: fix the typo

File: src/components/charts/modal/DeployChart.tsx [hunk 14]
```diff
@@ -350,20 +353,27 @@ const DeployChart: React.FC<DeployChartProps> = ({
         }
     }
 
-    async function handleDelete(force) {
+    async function handleDelete(deleteAction: string) {
```
Comment (vivek-devtron) [inline src/components/charts/modal/DeployChart.tsx:356]: change the type

File: src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx [hunk 15]
```diff
@@ -92,30 +98,64 @@ function EnvironmentSelectorComponent({
...
+        if (serverError instanceof ServerErrors && Array.isArray(serverError.errors)) {
+            serverError.errors.map(({ userMessage, internalMessage }) => {
+                setForceDeleteDialogTitle(userMessage)
+                setForceDeleteDialogMessage( internalMessage)
+            })
+        }
+    }
 
-    const getDeleteApplicationApi = (): Promise<any> => {
+    const getDeleteApplicationApi = (deleteAction: string): Promise<any> => {
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx:110]: change type

File: src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx [hunk 16]
```diff
@@ -92,30 +98,64 @@ function EnvironmentSelectorComponent({
...
+    const onClickHideNonCascadeDeletePopup = () => {
+        showNonCascadeDeleteDialog(false)
+    }
+    
+    const onClickNonCascadeDelete = async() => {
+        await deleteResourceAction(DELETE_ACTION.NONCASCADE_DELETE)
     }
 
-    async function deleteResourceAction() {
+    async function deleteResourceAction(deleteAction: string) {
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx:126]: same here

File: src/components/v2/values/chartValuesDiff/ChartValuesView.tsx [hunk 17]
```diff
@@ -562,34 +564,47 @@ function ChartValuesView({
         }
     }
 
-    const deleteApplication = (force?: boolean) => {
+    const deleteApplication = (deleteAction: string) => {
```
Comment (vivek-devtron) [inline src/components/v2/values/chartValuesDiff/ChartValuesView.tsx:567]: change type

File: src/components/v2/values/chartValuesDiff/ChartValuesView.tsx [hunk 18]
```diff
@@ -618,13 +646,13 @@ function ChartValuesView({
             })
     }
 
-    const getDeleteApplicationApi = (force?: boolean): Promise<any> => {
+    const getDeleteApplicationApi = (deleteAction: string): Promise<any> => {
```
Comment (vivek-devtron) [inline src/components/v2/values/chartValuesDiff/ChartValuesView.tsx:649]: same here

Comment (Ash-exp) [reply src/components/v2/values/chartValuesDiff/ChartValuesView.tsx]: fixed 👍

File: src/components/v2/values/chartValuesDiff/ChartValuesView.tsx (hunk 17 above)
Comment (Ash-exp) [reply src/components/v2/values/chartValuesDiff/ChartValuesView.tsx]: fixed 👍

File: src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx (hunk 16 above)
Comment (Ash-exp) [reply src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx]: fixed 👍

File: src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx (hunk 15 above)
Comment (Ash-exp) [reply src/components/v2/appDetails/sourceInfo/EnvironmentSelector.component.tsx]: fixed 👍

File: src/components/charts/modal/DeployChart.tsx (hunk 14 above)
Comment (Ash-exp) [reply src/components/charts/modal/DeployChart.tsx]: fixed 👍

File: src/components/charts/modal/DeployChart.tsx (hunk 13 above)
Comment (Ash-exp) [reply src/components/charts/modal/DeployChart.tsx]: fixed 👍

File: src/components/charts/discoverChartDetail/ChartDeploymentList.tsx (hunk 12 above)
Comment (Ash-exp) [reply src/components/charts/discoverChartDetail/ChartDeploymentList.tsx]: fixed 👍

File: src/components/cdPipeline/CDPipeline.tsx (hunk 11 above)
Comment (Ash-exp) [reply src/components/cdPipeline/CDPipeline.tsx]: fixed 👍

File: src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx (hunk 9 above)
Comment (Ash-exp) [reply src/components/common/ClusterNotReachableDailog/ClusterNotReachableDialog.tsx:13]: I handled the space in clusterNameString itself. clusterNameString : ` Cluster Name` || `NULL String`

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**